File: MiniDB_Projects/Team_23bcs10071/src/minidb/heapfile.py
import os
from .page import Page, PAGE_SIZE

class HeapFile:
    def __init__(self, path):
        self.path = path
        if not os.path.exists(path):
            # create empty file
            open(path, 'wb').close()

    def num_pages(self):
        size = os.path.getsize(self.path)
        return size // PAGE_SIZE

    def read_page(self, page_no):
        with open(self.path, 'rb') as f:
            f.seek(page_no * PAGE_SIZE)
            data = f.read(PAGE_SIZE)
            if len(data) < PAGE_SIZE:
                data = data + bytes(PAGE_SIZE - len(data))
            return Page(data)

    def append_page(self, page: Page):
        with open(self.path, 'ab') as f:
            f.write(page.data)
        return self.num_pages() - 1

    def write_page(self, page_no, page: Page):
        with open(self.path, 'r+b') as f:
            f.seek(page_no * PAGE_SIZE)
            f.write(page.data)

File: MiniDB_Projects/Team_IndexIQ/src/storage/heap_file.cpp
#include "heap_file.h"

HeapFile::HeapFile(const std::string& path) : path_(path) {
    file_.open(path, std::ios::in | std::ios::out | std::ios::binary);
    if (!file_.is_open()) {
        std::fstream create(path, std::ios::out | std::ios::binary);
        create.close();
        file_.open(path, std::ios::in | std::ios::out | std::ios::binary);
    }
}

HeapFile::~HeapFile() {
    if (file_.is_open()) file_.close();
}

Page HeapFile::read_page(uint32_t page_id) {
    Page p;
    p.id = page_id;
    file_.seekg(static_cast<std::streamoff>(page_id) * PAGE_SIZE);
    file_.read(reinterpret_cast<char*>(p.data), PAGE_SIZE);
    return p;
}

void HeapFile::write_page(const Page& page) {
    file_.seekp(static_cast<std::streamoff>(page.id) * PAGE_SIZE);
    file_.write(reinterpret_cast<const char*>(page.data), PAGE_SIZE);
    file_.flush();
}

uint32_t HeapFile::total_pages() {
    file_.seekg(0, std::ios::end);
    auto sz = file_.tellg();
    if (sz <= 0) return 0;
    return static_cast<uint32_t>(sz) / PAGE_SIZE;
}

uint32_t HeapFile::alloc_page() {
    uint32_t id = total_pages();
    uint8_t zeros[PAGE_SIZE]{};
    file_.seekp(static_cast<std::streamoff>(id) * PAGE_SIZE);
    file_.write(reinterpret_cast<const char*>(zeros), PAGE_SIZE);
    file_.flush();
    return id;
}
